pipeline-model-declarative-agent/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/agent/DeclarativeAgentDescriptor.java

```diff
@@ -0,0 +1,194 @@
+     * @return A map of descriptors with no arguments.
+     */
+    public static Map<String,DescribableModel> zeroArgModels() {
+        Map<String,DescribableModel> models = new HashMap<>();
+
+        for (Map.Entry<String,DescribableModel> entry : getDescribableModels().entrySet()) {
+            if (entry.getValue().getParameters().isEmpty()) {
+                models.put(entry.getKey(), entry.getValue());
+            }
+        }
+
+        return models;
+    }
+
+    /**
+     * An ordered (by Extension ordinal, in descending order) list of all descriptors.
+     * @return A list of descriptors.
+     */
+    public static List<DeclarativeAgentDescriptor> getOrderedDescriptors() {
+        List<DeclarativeAgentDescriptor> orderedDescriptors = new ArrayList<>();
+
+        List<ExtensionComponent<DeclarativeAgentDescriptor>> extensionComponents = new ArrayList<>(all().getComponents());
+        Collections.sort(extensionComponents);
+
+        for (ExtensionComponent<DeclarativeAgentDescriptor> extensionComponent: extensionComponents) {
+            orderedDescriptors.add(extensionComponent.getInstance());
+        }
+
+        return orderedDescriptors;
+    }
+
+    /**
+     * An ordered list of all descriptor names.
+     *
+     * @return A list of names
+     */
+    public static List<String> getOrderedNames() {
+        List<String> orderedNames = new ArrayList<>();
+
+        for (DeclarativeAgentDescriptor d : getOrderedDescriptors()) {
+            orderedNames.addAll(SymbolLookup.getSymbolValue(d));
+        }
+
+        return orderedNames;
+    }
+
+    /**
+     * Get the descriptor for a given name or null if not found.
+     *
+     * @param name The name for the descriptor to look up
+     * @return The corresponding descriptor or null if not found.
+     */
+    @Whitelisted
+    public static @Nullable DeclarativeAgentDescriptor byName(@Nonnull String name) {
+        return (DeclarativeAgentDescriptor) SymbolLookup.get().findDescriptor(DeclarativeAgent.class, name);
+    }
+
+    /**
+     * For a given name and map of arguments, find the corresponding descriptor and return an instance using those arguments.
+     *
+     * @param name The name of the descriptor
+     * @param arguments A map of arguments
+     * @return The instantiated {@link DeclarativeAgent} instance, or null if the name isn't found.
+     * @throws Exception
+     */
+    @Whitelisted
+    public static @Nullable DeclarativeAgent instanceForName(@Nonnull String name,
+                                                             Map<String,Object> arguments) throws Exception {
+        DeclarativeAgentDescriptor descriptor = byName(name);
+
+        if (descriptor != null) {
+            if (zeroArgModels().keySet().contains(name)) {
+                return descriptor.newInstance();
+            } else {
+                return descriptor.newInstance(arguments);
```

Comment:
`zeroArgModels` is a convenience method for finding, well, zero-arg models. The problem is that even in the case of zero-arg models, there's still an entry in the arg map - the key identifying the model. And for some reason I can't remember off the top of my head, even passing an empty `Map` still didn't work for instantiating them...

Oh, and `zeroArgModels` is also used at validation time to make sure any identifying-key-only calls are actually using a valid zero-arg model.